source/Model.cpp

```diff
@@ -1454,6 +1407,67 @@ Model Model::from_config_json(
   return model;
 }
 
+void Model::read_taint_configs_from_json(
+    const Json::Value& value,
+    std::vector<std::pair<AccessPath, Json::Value>>& generation_values,
+    std::vector<std::pair<AccessPath, Json::Value>>& parameter_source_values,
+    std::vector<std::pair<AccessPath, Json::Value>>& sink_values,
+    bool (*predicate)(const Json::Value& value)) {
+
+  for (auto generation_value :
+       JsonValidation::null_or_array(value, /* field */ "generations") |
+       boost::adaptors::filtered(predicate)) {
+    auto port = AccessPath(Root(Root::Kind::Return));
+    if (generation_value.isMember("port")) {
+      JsonValidation::string(generation_value, /* field */ "port");
+      port = AccessPath::from_json(generation_value["port"]);
+    } else if (generation_value.isMember("caller_port")) {
+      JsonValidation::string(generation_value, /* field */ "caller_port");
+      port = AccessPath::from_json(generation_value["caller_port"]);
+    }
+    generation_values.emplace_back(std::move(port), std::move(generation_value));
+  }
```

Comment:
I see why we want this helper but in general this seems clunky and error prone to use currently. 
All these loops are pretty much the same anyway. i.e: 
- Iterate on an array of json objects, filtering based on predicate
- Look for the "port" or "caller_port" or default to `"Root::Kind::Return"`
- Create an `AccessPath` from the port 
- Add mapping from AccessPath -> JsonValue.

So, instead can we turn this helper to operate on a single json array at a time and can call it for "generations"/"parameter_sources"/"sinks" separately. Callsite would instead be: 
```
auto generations_value = read_taint_configs_from_json(JsonValidation::null_or_array(value, "generations"), TaintConfigTemplate::is_concrete);
```

Reply:
I created a helper `read_taint_configs` for this purpse. It has a few more parameters, since the various loops did not behave exactly in the same way, e.g. w.r.t. defaulting to `Root::Kind::Return`. However, happy to adapt that some more if it is okay to change the user experience like that.

I kept `Model::read_taint_configs_from_json`, which now just invokes `read_taint_configs` with various parameters, as suggested by your example. However, since `Model::read_taint_configs_from_json` is invoked multiple times, I preferred to keep it rather than inlining it in multiple locations.